memcached.c

```diff
@@ -4191,6 +4287,837 @@ static inline void process_get_command(conn *c, token_t *tokens, size_t ntokens,
+                (it->exptime == 0) ? -1 : (current_time - it->exptime),
+                (unsigned long long)(current_time - it->time),
+                (unsigned long long)ITEM_get_cas(it),
+                (it->it_flags & ITEM_FETCHED) ? "yes" : "no",
+                ITEM_clsid(it),
+                (unsigned long) ITEM_ntotal(it));
+
+        item_remove(it);
+        c->wbytes = total + ret;
+        c->wcurr = c->wbuf;
+        conn_set_state(c, conn_write);
+        c->write_and_go = conn_new_cmd;
+    } else {
+        out_string(c, "EN");
+    }
+    pthread_mutex_lock(&c->thread->stats.mutex);
+    c->thread->stats.mget_cmds++;
+    pthread_mutex_unlock(&c->thread->stats.mutex);
+}
+
+#define MFLAG_MAX_OPT_LENGTH 20
+#define MFLAG_MAX_OPAQUE_LENGTH 32
+
+struct _meta_flags {
+    unsigned int no_update :1;
+    unsigned int locked :1;
+    unsigned int vivify :1;
+    unsigned int la :1;
+    unsigned int hit :1;
+    unsigned int value :1;
+    unsigned int set_stale :1;
+    unsigned int no_reply :1;
+};
+
+static int _meta_flag_preparse(char *opts, size_t olen, struct _meta_flags *of) {
+    unsigned int i;
+    // NOTE: 'seen' is one of those need-to-microbench situation to do this via bit vector or
+    // simple 8 bit array; need two divs and a shift vs a bit more memory.
+    // Leave it simple for now, optimize later.
+    uint8_t seen[127] = {0};
+    // also count how many tokens should be necessary to parse.
+    int tokens = 0;
+    for (i = 0; i < olen; i++) {
+        uint8_t o = (uint8_t)opts[i];
+        // zero out repeat flags so we don't over-parse for return data.
+        if (o >= 127 || seen[o] != 0) {
+            opts[i] = 0;
+            continue;
+        }
+        seen[o] = 1;
+        // FIXME: alphabetize.
+        switch (opts[i]) {
+            case 'N':
+                of->locked = 1;
+                of->vivify = 1;
+                tokens++;
+                break;
+            case 'T':
+                tokens++;
+                of->locked = 1;
+                break;
+            case 'R':
+                of->locked = 1;
+                tokens++;
+                break;
+            case 'l':
+                of->la = 1;
+                of->locked = 1; // need locked to delay LRU bump
+                break;
+            case 'O':
+                tokens++;
+                break;
+            case 'k':
+                break;
+            case 'h':
+                of->locked = 1; // need locked to delay LRU bump
+                break;
+            case 'u':
+                of->no_update = 1;
+                break;
+            case 'q':
+                of->no_reply = 1;
+                break;
+            // mset-related.
+            case 'F':
+            case 'S':
+            case 'C':
+                tokens++;
+                break;
+            case 'I':
+                of->set_stale = 1;
+                break;
```

Comment:
Shouldn't we error out if user specifies at least one wrong option?

Reply:
Not honestly sure. I'd rather have the parser shared between all of the code, so bad flags are just ignored. Else you'd have to map everything out which is a lot more code.

Reply:
`tokens` should be always >=0, can we use `-1` as the return code to indicate a parser error?

Reply:
Not sure I follow? You still need to know which flags pertain to the sub-command in the parser here. That might make the code too obtuse.